File: MiniMdb.Backend/Shared/ApiMessage.cs
namespace MiniMdb.Backend.Shared
{
    /// <summary>
    /// Unified envelope for API messages.
    /// </summary>
    public class ApiMessage
    {
        /// <summary>
        /// Not null when API call fails
        /// </summary>
        public ApiError Error { get; set; }

        /// <summary>
        /// Not null when API method supports pagination
        /// </summary>
        public ApiPagination Pagination { get; set; }

        public static ApiMessage MakeError(int code, string message)
        {
            return new ApiMessage { Error = new ApiError { Code = code, Message = message } };
        }

        public static ApiMessage<T> From<T>(ApiPagination pagination, params T[] data)
        {
            return new ApiMessage<T> { Data = data, Pagination = pagination };
        }

        public static ApiMessage<T> From<T>(T data)
        {
            return new ApiMessage<T> { Data = new[] { data } };
        }
    }

    /// <summary>
    /// Unified envelope for API result with data.
    /// </summary>
    /// <typeparam name="T">data type of successful execution</typeparam>
    public class ApiMessage<T> : ApiMessage
    {
        /// <summary>
        /// API result in case of successful execution
        /// </summary>
        public T[] Data { get; set; }
    }
}

File: MiniMdb.Backend/ClientApp/src/app/models/ApiMessage.ts
/**
 * Base API message envelope
 */
export class ApiMessageBase {
  error: ApiError;
  pagination: ApiPagination;
}

/**
 * API message envelope with data
 */
export class ApiMessage<T> extends ApiMessageBase {
  data: T[];
}

/**
 * API error description
 */
export class ApiError {
  code: number;
  message: string;
}

/**
 * Pagination details
 */
export class ApiPagination {
  count: number;
  page: number;
  pageSize: number;
  totalPages: number;
}
